### ISSUE TITLE
Two-step push to work around `git push` after LFS migration on big repo seemingly stuck?

### ISSUE THREAD

**Issue by Skymirrh** [git-lfs/git-lfs#4350]
# The issue
The first `git push` after an LFS migration on a big repository may take several hours or days to complete, seemingly stuck while uploading LFS objects:

```console
$ git push
Uploading LFS objects: 100% (2312/2312), 765 MB | 4.6 MB/s
```

In my case, "big" means (on a fresh `git clone --mirror`):
- 1.5&nbsp;GB bare total size (`du -sh .`)
- 700K Git objects (`git rev-list --objects --all | wc -l`)
- Around 2600 LFS objects (`find lfs/objects/ -type f | wc -l`) after a `git lfs migrate import --everything --include="*.ext1,*.ext2"` pass

# Analysis

Enabling debug output yields more detail:

```console
$ GIT_TRACE=1 git push
Uploading LFS objects: 100% (2312/2312), 765 MB | 4.6 MB/s
16:41:13.318312 trace git-lfs: tq: running as batched queue, batch size of 100
16:41:13.318312 trace git-lfs: run_command: git rev-list --stdin --objects --not --remotes=<redacted> --
... [repeating]
```

A quick code inspection reveals [this log happens while processing the transfer queue](https://github.com/git-lfs/git-lfs/blob/v2.13.1/tq/transfer_queue.go#L979), and helped me find existing issues with a similar behavior ([#3707](https://github.com/git-lfs/git-lfs/issues/3707), [#3915](https://github.com/git-lfs/git-lfs/issues/3915)), and where the following is stated:

> [...] we're traversing the entire history to see what objects need to be pushed. If you have a large repository with a large history, that can be expensive. It's unfortunately unavoidable, since we need to know which objects have to be pushed, and the only way to know that is to read all of the pointer files from all of the unpushed history.

> When we're pushing a repository that's newly rewritten, we're going to traverse every blob that exists that's less than 1024 bytes so that we can determine if an LFS object is in that object.

This explains why `git push` takes a very long time processing LFS objects on this repo, if it has to go through all 700K Git objects, 100 by 100, to determine whether or not they are LFS objects.

Note: the exact same issue happens when trying to push all LFS objects with `git lfs push --all`, which is coherent as it seems to basically be what `git push` is doing thanks to the hooks installed via `git lfs install`.

# Improvising a two-step push

I tried to leave `git push` doing its thing, but after 24 hours I decided it was unsuitable for us, especially since it was clear we were nowhere near the end: the object count was around 2500, whereas it started around 2300 and I knew there was a total of 2600 LFS objects to upload (`find lfs/objects/ -type f | wc -l`).

This last bit was key to a potential solution: since I could list all LFS objects, I wondered if I could split the push operation in two steps, one for regular Git objects and the other for LFS objects. This way we would avoid the need to scan through all blobs.

Taking advantage of [`git lfs push`'s `--object-id` option](https://github.com/git-lfs/git-lfs/blob/v2.13.1/docs/man/git-lfs-push.1.ronn), I seemingly successfully did a two-step push:

```shell
git lfs uninstall # disable LFS hooks
git lfs push origin --object-id `find lfs/objects/ -type f -printf "%f "` # push LFS objects
git push # push regular Git objects
git lfs install # re-enable LFS hooks
```

This goes as fast as network speed allows. We can then try a regular `git push` again, and receive `Everything up-to-date` as a kinda confirmation that everything seemingly worked. And indeed, subsequent clones and regular Git operations with the migrated repo are seemingly working perfectly fine.

# So my question is: what am I missing?

I must be missing something, otherwise this would be the default behavior, no scanning involved 🤔

**Comment by bk2204** [comment]
Hey,

This is a great question.  The reason that your solution is efficient in your case is because you know that the remote repository has no LFS objects and you'll need to upload them all.  However, when we use the `pre-push` hook to do an upload, we have no way of knowing what refs may already exist on the remote side.  All we know is that none of the refs being pushed exist yet.  This case, generally, is indistinguishable from pushing a small number of new branches with only a few commits.

In the case of an LFS migration into an existing repository, you just look like you're updating many refs at once, and that could be just a single commit on each one.  We don't know until we start to look at history.

In the general case, walking history is much, much faster than pushing all objects, _especially_ on slow connections, because we'll traverse a relatively small number of commits.  You're correct that in this case it is much slower, but we generally recommend for a new push that people do this:

```sh
$ git lfs push --all
$ git push --mirror --no-verify
```

That's because you do definitely want to push all of your LFS objects when you're doing any sort of migration, but, having uploaded them, you don't need to walk history.  This will be much faster and is equivalent to your solution.

**Comment by Skymirrh** [comment]
Thanks for the thorough explanation and recommendation. Bypassing the `pre-push` hook at `push` time is a neat trick, didn't think of that!

I'm actually a bit puzzled though, as I thought I was experiencing the same issue when using `git lfs push --all`, as denoted above. I probably have made a mistake while analyzing, but just to be sure I'm going to try it on another repo, as I have kept intermediate backup copies of my various tries... :D

**Comment by bk2204** [comment]
Hmm,

It may be that we do traverse history in that case as well.  I would argue that's probably not necessary and we could improve there, since we know that we're pushing literally every object in the repository.  Let me mark this as an enhancement for that reason.

**Comment by Skymirrh** [comment]
Just tried again: can confirm that from bare mirror clone with `origin` set up as mirror remote (`git config remote.origin.mirror` is `true`), `git lfs push origin --all` seems to be traversing history in the same way as `git push` does.

Also, I was thinking maybe I could document the workaround in https://github.com/git-lfs/git-lfs/wiki/Tutorial#migrating-existing-repository-data-to-lfs, if that sounds good to you? This is the place I looked at first when doing my own migration.

**Comment by Skymirrh** [comment]
Looks like the new flag `GIT_LFS_SKIP_PUSH` may also be a better option that `--no-verify`, so as to allow potential other `pre-push` hooks to still do their office.

**Comment by bk2204** [comment]
Yeah, I think that updating the tutorial would be great.  Since we'd be using `GIT_LFS_SKIP_PUSH`, that would also obviate the need for running `git lfs uninstall` and `git lfs install`.

**Comment by Skymirrh** [comment]
Cool, I'll make a change proposal when I have some time, probably after December 30. Happy holidays!

**Comment by nicolasmanic** [comment]
I have the same issue as above, I believe removing the craw of history for `git lfs push --all` is a correct step. 
In my case I have the following log messaging every time I initiate the cmd:
```
trace git-lfs: tq: running as batched queue, batch size of 100
trace git-lfs: run_command: git rev-list --objects --do-walk --stdin --
trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'cat-file' '--batch-check'
trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'rev-parse' '--git-common-dir'
.
.
.
race git-lfs: tq: running as batched queue, batch size of 100
trace git-lfs: run_command: git rev-list --objects --do-walk --stdin --
trace git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'cat-file' '--batch-check'
race git-lfs: exec: git '-c' 'filter.lfs.smudge=' '-c' 'filter.lfs.clean=' '-c' 'filter.lfs.process=' '-c' 'filter.lfs.required=false' 'rev-parse' '--git-common-dir'
trace git-lfs: tq: running as batched queue, batch size of 100
```

Also the workaround suggested above will not work in my case since we have ~40.000 lfs objects

**Comment by nbusseneau** [comment]
> Also the workaround suggested above will not work in my case since we have ~40.000 lfs objects

Curious as to why it will not work. I had around 2.6K objects, so not close to 40K, but I don't see why it wouldn't scale considering the simplicity of the workaround 🤔

**Comment by nicolasmanic** [comment]
There is a limit on how many object id's you can pass to  `git lfs push <remote> --object-id` and it's less than 40.000. I guess I can create something smarter and do 40.000 `git lfs push <remote> --object-id`  for each object instead of one with 40.000 object IDs.

**Comment by nbusseneau** [comment]
Or just split your 40K objects in chunks, the `git lfs push origin --object-id <ids>` does not need to push all objects in one sweep ;)

**Comment by triplef** [comment]
Has there been new developments on this issue? It looks like the official documentation does not yet mention any ways to improve initial pushes of large repositories.

I’m trying to migrate a large repository from another provider to GitHub. It has about 40k LFS objects, and doing a regular push takes over a day even on an EC2 instance with fast networking due to git crawling the history.

I tried running `git lfs push --all` first as suggested [above](https://github.com/git-lfs/git-lfs/issues/4350#issuecomment-749036247), but that seems to also crawl the history.

I’m now trying to push individual LFS objects based on @nbusseneau’s suggestions, but due to the large number of objects using xargs to split the calls:
```shell
find lfs/objects/ -type f -printf "%f\0" | xargs -0 git lfs push origin --object-id
git push --mirror --no-verify
```

(I’m using `--no-verify` instead of `GIT_LFS_SKIP_PUSH=1` because the machine has Git LFS 2.12.1.)

Is that still the recommended approach? Would be great to see this documented somewhere, or ideally have a flag like `git push --mirror --initial-push` added that does the same internally.

**Comment by nbusseneau** [comment]
> Has there been new developments on this issue? It looks like the official documentation does not yet mention any ways to improve initial pushes of large repositories.

Yeah sorry, my bad for that one, I said I would try to submit doc changes but never did.

> Is that still the recommended approach?

Seeing as there were no updates on this issue I believe this should still be the way to go: it would make sense to me this issue would have seen updates or been closed if anything had developed.

**Comment by awkto** [comment]
I am also running into this issue. Pushing LFS objects to a fresh repo hangs at 100% after the `git lfs push --all`

Since I can't pass `--no-verify` to the `git lfs push --all`, is there any solution or workaround for this initial lfs push?

**Comment by ak-ldube** [comment]
Thank you for the investigation! This worked flawlessly on our side. We have a very large repository that we are migrating to LFS and the push was driving me crazy and not finishing after a few days. This made the push take minutes instead.
